components/config/src/theme.rs

```diff
@@ -2,40 +2,55 @@ use std::collections::HashMap;
-use toml::Value as Toml;
+use unic_langid::LanguageIdentifier;
 
-use errors::{bail, Result};
+use crate::config::languages::LocaleOptions;
+use errors::Result;
 use utils::fs::read_file_with_error;
 
 /// Holds the data from a `theme.toml` file.
-/// There are other fields than `extra` in it but Zola
-/// itself doesn't care about them.
-#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+///
+/// There are other fields, but Zola only cares about `[extra]` and `[languages]`.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(default)]
 pub struct Theme {
-    /// All user params set in [extra] in the theme.toml
-    pub extra: HashMap<String, Toml>,
+    /// The primary/default language of the theme
+    ///
+    /// MUST be a valid language, specified using the [BCP 47] syntax. To retain backwards
+    /// compatibility with other language naming systems, the [language_alias] option was added.
+    ///
+    /// If not specified, [default_language_options] will be merged into
+    /// [config.default_language_options] regardless of [config.default_language].
+    pub default_language: Option<LanguageIdentifier>,
+    /// Theme-wide defaults for locale-specific options
+    ///
+    /// All values except those in `[extra]` are ignored.
+    ///
+    /// This is flattened when serializing/deserializing, i.e. its fields will appear as if they
+    /// were at the top level alongside base_url. These values are used for the default language,
+    /// and other languages will fall back to these values if they are not overridden in
+    /// `languages`
+    #[serde(flatten, default = "LocaleOptions::default_for_default_lang")]
+    pub default_language_options: LocaleOptions,
+
+    /// Language-specific options for translations of this site
+    ///
+    /// All values except those in `[extra]` are ignored.
+    ///
+    /// The key of the TOML table (represented as a HashMap in Rust) is a valid language code, and
+    /// its values can be used for overriding some of the site-wide settings.
+    ///
+    /// See [LocaleOptions] for what options can be specified here and what their default values
+    /// are.
+    ///
+    /// Having an entry for the default language is treated as an error.
+    pub languages: HashMap<LanguageIdentifier, LocaleOptions>,
 }
 
 impl Theme {
     /// Parses a TOML string to our Theme struct
     pub fn parse(content: &str) -> Result<Theme> {
-        let theme = match content.parse::<Toml>() {
-            Ok(t) => t,
-            Err(e) => bail!(e),
-        };
-
-        let mut extra = HashMap::new();
-        if let Some(theme_table) = theme.as_table() {
-            if let Some(ex) = theme_table.get("extra") {
-                if ex.is_table() {
-                    extra = ex.clone().try_into().unwrap();
-                }
-            }
-        } else {
-            bail!("Expected the `theme.toml` to be a TOML table")
-        }
-
-        Ok(Theme { extra })
```

Comment:
how come we don't need that anymore?

Reply:
The code previously parsed the toml into  a Toml object and just copied the `extra` key from it into the struct. This did not use Serde at all.

The new code just uses Serde to directly parse that into the `Theme` struct.

I should probably look up if an attribute needs to be set to ignore unknown keys.